### ISSUE TITLE
ENH: Add possibility to optimize memory usage

### ISSUE THREAD

**Issue by HTenkanen** [HTenkanen/pyrosm#87]
Pyrosm parses quite a lot of attributes by default from the OSM data. Add possibility to optimize memory usage by keeping only the most necessary attributes in the results and dropping everything else out. This helps when working with large files. 

The memory optimization should influence the behavior when parsing the PBF (i.e. not parsing less important attributes at all) and when parsing the tags (keeping only the most relevant ones).

**Comment by chourmo** [comment]
I was going to create an enhancement request to add a "minimal" network_type keyword to get_network, with only geometry and id as keys. Does it fit in this issue ?

**Comment by HTenkanen** [comment]
@chourmo Yes, that is more or less the idea with this issue 🙂 There should be a parameter to reduce the number of attributes to absolute minimum, and in addition to that give the user the possibility to keep some specific tag using the `tags_to_keep` parameter or something like that.

**Comment by HTenkanen** [comment]
Implementation ideas for this:
- OSM() should have a parameter `keep_meta` that could be used to skip reading metadata from PBF (version, timestamp, etc.)
- all "get" methods should have parameter `optimize_memory` that can be set to True for optimizing memory usage by parsing only crucial attributes from the data (e.g. id and geometry)
- all get methods should have possibility to specify which are those crucial attributes by passing a list of names (`tags_to_keep`?)
- to_graph() should have parameters `keep_edge_geometry` and `keep_node_geometry` that could be used to drop geometry attribute from the graph which are not necessarily needed always

**Comment by chourmo** [comment]
"optimize memory" seems quite ambiguous to me, as the impact on tags is not obvious. "minimal_tags" seems to express results more clearly.

**Comment by HTenkanen** [comment]
@chourmo Thanks! That is indeed more intuitive option 👍

**Comment by chourmo** [comment]
It would also be create great to have geometry=True/False, default to True, to only retrieve tags.

**Comment by HTenkanen** [comment]
> It would also be great to have geometry=True/False, default to True, to only retrieve tags.

@chourmo The idea is to allow user to have full control with `tags_to_keep` parameter. So using that you can specify e.g. tags_to_keep=["id", "name"] which would only keep those ones, i.e. meaning that geometry would be dropped out from the final result. The only required tag that will always be kept is the `id` (due to how the internals of the library work).

**Comment by HTenkanen** [comment]
Take advantage of `sparse` arrays? Would introduce a new dependency ...
https://sparse.pydata.org/en/stable/

**Comment by tfardet** [comment]
Not sure whether it's the right place to post this, but there is a major possibility for memory improvement when using ``bounding_box`` in ``pyrosm.OSM``.

A naive implementation I tested was to move [nodes parsing](https://github.com/HTenkanen/pyrosm/blob/master/pyrosm/pbfreader.pyx#L393) directly into the [file reading loop](https://github.com/HTenkanen/pyrosm/blob/master/pyrosm/pbfreader.pyx#L55), basically deleting ``get_primitive_blocks_and_string_tables`` and moving its code directly into ``_parse_osm_data``.

On a specific example with the osm.pbf of a French region where I'm only interested in one city, I get the memory footprint of ``get_buildings`` down from 8 GB to 200 MB, so a 4000% decrease.
The downside is that speed goes down from a bit less than 40 s to more than 60 s (close to 200% increase), but there might be ways to improve that.

**Comment by chungkang** [comment]
I had same issue.
I tried to use osmium tool to crop the data in small region with osmium-extract.
https://docs.osmcode.org/osmium/latest/osmium-extract.html
I could crop and read the pbf file, but still cannot process it with pyrosm.

**Comment by tfardet** [comment]
@chungkang I ended up combining ``osmium`` and ``[pyogrio](https://pyogrio.readthedocs.io)``, so switching away from ``pyrosm`` as this new library is actively developed by the people from ``geopandas``.

**Comment by HTenkanen** [comment]
Memory optimization has landed for the upcoming **v0.9.0** 🙂 New `OSM(..., keep_metadata=False)` drops element metadata, `tags_to_keep` keeps only the tag-columns you choose, and the read path now streams blocks and builds only the columns that actually occur — cutting peak memory ~20% on a whole-file read and ~50–70% for bounding-box reads. Thanks @chourmo for the concrete API input (the metadata/column-selection direction), and @tfardet and @chungkang for the extra information and workarounds. Merged to `master`; closing as fixed. 🙏

**Comment by tfardet** [comment]
Thanks @HTenkanen 
From your message, it's a bit unclear for me what leads to memory improvement with a bbox... (I don't know my way around the codebase enough to find it: only looked at it once, 3 years ago)
Is it only if ``keep_metadata`` is set to false or does it occur whatever happens, a bit like what I did during my tests?

**Comment by HTenkanen** [comment]
Hi @tfardet, there are actually quite many things updated to improve and control the memory consumption. The most important one is streaming block reads that reduces memory usage significantly especially if bounding box is applied. This document describes more all different approaches used to decrease the peak memory usage (the actual implementations you can check from recent PRs): https://github.com/pyrosm/pyrosm/blob/master/docs/benchmarks/memory_results.md
